src/main/java/gregtech/common/pipelike/laser/net/LaserPipeNet.java

```diff
@@ -0,0 +1,120 @@
+package gregtech.common.pipelike.laser.net;
+
+import gregtech.api.capability.GregtechTileCapabilities;
+import gregtech.api.capability.ILaserContainer;
+import gregtech.api.pipenet.Node;
+import gregtech.api.pipenet.PipeNet;
+import gregtech.api.pipenet.WorldPipeNet;
+import gregtech.api.util.FacingPos;
+import gregtech.common.pipelike.laser.LaserPipeProperties;
+import it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap;
+import net.minecraft.nbt.NBTTagCompound;
+import net.minecraft.tileentity.TileEntity;
+import net.minecraft.util.EnumFacing;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.world.World;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.Map;
+
+public class LaserPipeNet extends PipeNet<LaserPipeProperties> {
+
+    private final Map<BlockPos, LaserData> NET_DATA = new Object2ObjectOpenHashMap<>();
+
+    public LaserPipeNet(WorldPipeNet<LaserPipeProperties, ? extends PipeNet<LaserPipeProperties>> world) {
+        super(world);
+    }
+
+    @Nullable
+    public LaserData getNetData(BlockPos pipePos, EnumFacing facing) {
+        LaserData data = NET_DATA.get(pipePos);
+        if (data == null) {
+            data = LaserNetWalker.createNetData(getWorldData(), pipePos, facing);
+            if (data == null) {
+                // walker failed, don't cache, so it tries again on next insertion
+                return null;
+            }
+
+            NET_DATA.put(pipePos, data);
+        }
+        return data;
+    }
+
+    @Override
+    public void onNeighbourUpdate(BlockPos fromPos) {
+        NET_DATA.clear();
+    }
+
+    @Override
+    public void onPipeConnectionsUpdate() {
+        NET_DATA.clear();
+    }
+
+    @Override
+    protected void transferNodeData(Map<BlockPos, Node<LaserPipeProperties>> transferredNodes, PipeNet<LaserPipeProperties> parentNet) {
+        super.transferNodeData(transferredNodes, parentNet);
+        NET_DATA.clear();
+        ((LaserPipeNet) parentNet).NET_DATA.clear();
+    }
+
+    @Override
+    protected void writeNodeData(LaserPipeProperties nodeData, NBTTagCompound tagCompound) {
+
+    }
+
+    @Override
+    protected LaserPipeProperties readNodeData(NBTTagCompound tagCompound) {
+        return new LaserPipeProperties();
+    }
+
+    // jabel moment
+    public static class LaserData {
```

Comment:
Nullity annotations and some javadocs for methods in this class would be helpful.